### PR TITLE
add optional component parameter for infrastructure

### PR DESCRIPTION
**Pull request by chris-pateman**
Add the optional parameter component to infrastructure pipelines.
When having more then one pipeline that only deploys infrastructure they clash as they use the same state file using the product name. This will give the option to pass in the component name, which is used in the state file location name.

### CODE REVIEW

File: vars/withInfraPipeline.groovy [hunk 1]
```diff
@@ -7,7 +7,7 @@ import uk.gov.hmcts.contino.Subscription
 import uk.gov.hmcts.contino.Environment
 import uk.gov.hmcts.pipeline.TeamConfig
 
-def call(String product, Closure body) {
+def call(String product, String component = "", Closure body) {
```
Comment (timja) [inline vars/withInfraPipeline.groovy:10]: use null not empty string
```suggestion
def call(String product, String component = null, Closure body) {
```

File: README.md [hunk 2]
```diff
@@ -266,7 +266,10 @@ Example `Jenkinsfile` to use the opinionated infrastructure pipeline:
 
 def product = "rhubarb"
 
-withInfraPipeline(product) {
+//Optional
+def component = "extra-detail"
```
Comment (timja) [inline README.md:270]: this should be a separate example, people tend to copy paste and given it's not our default approach then it shouldn't be included in this example

Comment (chris-pateman) [reply README.md:294]: I have added a new section. please review

File: vars/withInfraPipeline.groovy (hunk 1 above)
Comment (chris-pateman) [reply vars/withInfraPipeline.groovy]: updated

File: README.md [hunk 3]
```diff
@@ -264,7 +264,7 @@ Example `Jenkinsfile` to use the opinionated infrastructure pipeline:
 
 @Library("Infrastructure") _
 
-def product = "rhubarb"
+def product = "rhubarb"
```
Comment (timja) [inline README.md:267]:
```suggestion
def product = "rhubarb"
```

File: README.md [hunk 4]
```diff
@@ -273,6 +273,33 @@ withInfraPipeline(product) {
...
 ```
 
+#### Optional parameters for the opinionated infratructure pipeline
+
+You have the ability to pass extra parameters to the `withInfraPipeline`.
+
+These parameters include:
+| parameter name | description | example |
+| --- | --- | --- |
+| component | Extra detail to the name of the repository | "extra-detail" |
```
Comment (timja) [inline README.md:283]: not the best name really for the description, wondering if should just link to the glossary?
https://hmcts.github.io/glossary/#component

**Review by timja — DISMISSED** [review]

**Review by timja — APPROVED** [review]

**Pull request merged**